File: terminal-frontend/test/utils/age_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:clubbar_terminal/utils/age.dart';

/// The arithmetic the whole Jugendschutz gate rests on (epic #582 M6, ADR-0045).
///
/// It is pulled out as a pure function for one reason: **the birthday boundary
/// is where off-by-one errors live**, and it is far easier to state every case
/// against a fixed clock here than to reach them through a cart.
///
/// The terminal computes this itself, from its own clock, because it may not
/// have spoken to the server in weeks (decision 1). Nothing here reads a stored
/// age — an age in years is wrong from the member's next birthday onward.
void main() {
  group('hasReachedAge', () {
    // A fixed "today" so these cases mean the same thing next year.
    final today = DateTime(2026, 8, 20);

    test('the day before the birthday is still too young', () {
      expect(hasReachedAge(DateTime(2008, 8, 21), 18, today), isFalse);
    });

    /// The case an off-by-one gets wrong, and the one a member notices: they
    /// are 18 *on* their eighteenth birthday, not the day after.
    test('the birthday itself counts', () {
      expect(hasReachedAge(DateTime(2008, 8, 20), 18, today), isTrue);
    });

    test('the day after is obviously fine', () {
      expect(hasReachedAge(DateTime(2008, 8, 19), 18, today), isTrue);
    });

    test('a comfortably older member passes', () {
      expect(hasReachedAge(DateTime(1985, 6, 15), 18, today), isTrue);
    });

    test('a child fails both thresholds', () {
      final born2012 = DateTime(2012, 1, 1);
      expect(hasReachedAge(born2012, 16, today), isFalse);
      expect(hasReachedAge(born2012, 18, today), isFalse);
    });

    test('the two JuSchG thresholds are decided independently', () {
      // Seventeen: old enough for beer, not for spirits. The one member for
      // whom § 9's two limits actually differ.
      final seventeen = DateTime(2009, 3, 4);
      expect(hasReachedAge(seventeen, 16, today), isTrue);
      expect(hasReachedAge(seventeen, 18, today), isFalse);
    });

    /// A leap-day birth has no anniversary in three years out of four. German
    /// practice (§§ 187–188 BGB) puts the birthday at 1 March in those years,
    /// which is what `DateTime(2026, 2, 29)` normalizes to — so the rule is
    /// followed by the arithmetic rather than by a special case.
    group('a leap-day birth', () {
      final born = DateTime(2008, 2, 29);

      test('is not yet 18 on 28 February of a non-leap year', () {
        expect(hasReachedAge(born, 18, DateTime(2026, 2, 28)), isFalse);
      });

      test('turns 18 on 1 March of a non-leap year', () {
        expect(hasReachedAge(born, 18, DateTime(2026, 3, 1)), isTrue);
      });

      test('turns 18 on 29 February when the year has one', () {
        expect(hasReachedAge(born, 18, DateTime(2028, 2, 29)), isTrue);
      });
    });

    /// The time of day must not decide a legal question. A terminal open past
    /// midnight is the ordinary case for a bar.
    test('the clock time within the day is irrelevant', () {
      final born = DateTime(2008, 8, 20);
      expect(hasReachedAge(born, 18, DateTime(2026, 8, 20, 0, 0, 1)), isTrue);
      expect(hasReachedAge(born, 18, DateTime(2026, 8, 20, 23, 59, 59)), isTrue);
      expect(hasReachedAge(born, 18, DateTime(2026, 8, 19, 23, 59, 59)), isFalse);
    });

    test('a birth date in the future is never old enough', () {
      expect(hasReachedAge(DateTime(2027, 1, 1), 1, today), isFalse);
    });
  });

  group('parseCachedDateOfBirth', () {
    test('reads the wire format the sync stores', () {
      expect(parseCachedDateOfBirth('1985-06-15'), DateTime(1985, 6, 15));
    });

    test('reads a full timestamp, keeping only the day', () {
      expect(
        parseCachedDateOfBirth('1985-06-15T00:00:00.000Z'),
        DateTime(1985, 6, 15),
      );
    });

    /// Null is not "unknown" — it is an anonymized member (ADR-0045 rule 3),
    /// and the caller must refuse rather than fall open. Returning null here
    /// keeps that decision at the call site instead of inventing a date.
    test('gives nothing back for an absent or unusable value', () {
      expect(parseCachedDateOfBirth(null), isNull);
      expect(parseCachedDateOfBirth(''), isNull);
      expect(parseCachedDateOfBirth('not a date'), isNull);
    });
  });
}

File: backend/tests/Unit/Shared/Utils/AgeTest.php
<?php

declare(strict_types=1);

namespace Tests\Unit\Shared\Utils;

use App\Shared\Utils\Age;
use PHPUnit\Framework\Attributes\DataProvider;
use PHPUnit\Framework\TestCase;

/**
 * How old somebody was on a given day (epic #582 M7, ADR-0045).
 *
 * The server recomputes this at **the transaction's own timestamp**, not at
 * ingest time (ADR-0045 rule 2). A terminal may sync days or weeks after the
 * sale, and the question a Jugendschutz check answers is whether the sale was
 * lawful *when it happened* — not whether the member has since had a birthday.
 *
 * Arithmetic worth its own tests because the whole detection rests on it, and
 * because the boundary is where this class of bug lives: somebody who turns 18
 * on the day of the sale is 18, and a check that says 17 refuses a lawful drink
 * — or, the same error the other way, records a violation that did not happen.
 */
class AgeTest extends TestCase
{
    private static function at(string $iso): \DateTimeImmutable
    {
        return new \DateTimeImmutable($iso);
    }

    /** @return array<string, array{0: string, 1: string, 2: int}> */
    public static function ages(): array
    {
        return [
            // birth date, moment, expected full years
            'the day before the eighteenth birthday' => ['2008-06-15', '2026-06-14 20:00:00', 17],
            'the eighteenth birthday itself' => ['2008-06-15', '2026-06-15 20:00:00', 18],
            'one minute past midnight on the birthday' => ['2008-06-15', '2026-06-15 00:01:00', 18],
            'the day after' => ['2008-06-15', '2026-06-16 20:00:00', 18],
            'comfortably adult' => ['1985-06-15', '2026-08-20 12:00:00', 41],
            'born today' => ['2026-08-20', '2026-08-20 12:00:00', 0],
            'the day before a birthday, a month out' => ['2010-12-31', '2026-12-30 23:59:59', 15],
            'that same birthday' => ['2010-12-31', '2026-12-31 00:00:00', 16],
        ];
    }

    #[DataProvider('ages')]
    public function test_full_years_between_a_birth_date_and_a_moment(
        string $dateOfBirth,
        string $moment,
        int $expected,
    ): void {
        $this->assertSame($expected, Age::yearsAt($dateOfBirth, self::at($moment)));
    }

    /**
     * 29 February is the one birth date the calendar does not repeat every year.
     *
     * PHP's own diff answers this the way German law does — a leapling has their
     * birthday on 1 March in a common year (§ 188 Abs. 3 BGB reasoning) — so on
     * 28 February they are still a year younger. Pinned because a hand-rolled
     * month/day comparison typically gets it wrong in the other direction.
     */
    public function test_a_leap_day_birth_date(): void
    {
        $this->assertSame(17, Age::yearsAt('2008-02-29', self::at('2026-02-28 12:00:00')));
        $this->assertSame(18, Age::yearsAt('2008-02-29', self::at('2026-03-01 12:00:00')));
    }

    /**
     * The sale is judged on the day it happened, not on the day the batch
     * arrives. This is the whole reason the check runs on `occurred_at`.
     */
    public function test_the_moment_is_what_decides(): void
    {
        $bornSixteenOn20260615 = '2010-06-15';

        // Sold the day before the birthday, synced a fortnight later.
        $this->assertSame(15, Age::yearsAt($bornSixteenOn20260615, self::at('2026-06-14 21:30:00')));
        // The same member, a day later, is old enough.
        $this->assertSame(16, Age::yearsAt($bornSixteenOn20260615, self::at('2026-06-15 21:30:00')));
    }

    /** A timestamp with a zone is read as the instant it names. */
    public function test_an_iso_timestamp_with_a_zone(): void
    {
        $this->assertSame(18, Age::yearsAt('2008-06-15', self::at('2026-06-15T18:00:00Z')));
    }

    /** @return array<string, array{0: string}> */
    public static function unusableBirthDates(): array
    {
        return [
            'empty' => [''],
            'prose' => ['sometime in the eighties'],
            'a non-existent day' => ['1990-02-30'],
        ];
    }

    /**
     * Null rather than a guess.
     *
     * The caller has to decide what an unknowable age means, and for the server
     * that is "record nothing" — asserting a violation from a date it cannot
     * read would be inventing a legal finding out of missing data.
     */
    #[DataProvider('unusableBirthDates')]
    public function test_an_unreadable_birth_date_yields_null(string $dateOfBirth): void
    {
        $this->assertNull(Age::yearsAt($dateOfBirth, self::at('2026-08-20 12:00:00')));
    }

    /**
     * A birth date after the moment is not a negative age — it is nonsense, and
     * nonsense is null. `past_date` (M2) keeps this out of the database on the
     * write path; a stored row predating that rule, or a transaction backdated
     * by a terminal with a wrong clock, could still reach here.
     */
    public function test_a_birth_date_after_the_moment_yields_null(): void
    {
        $this->assertNull(Age::yearsAt('2026-12-31', self::at('2026-08-20 12:00:00')));
    }
}
